File: ios/Engine.swift
//
//  Engine.swift
//  Glimpse
//
//  Created by Nitin Issac Joy on 09/07/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import Photos
import RealmSwift

class Engine {

    static var EngineInstance = Engine()

    init() {
        //self.ExtendHomeDataToDate()
    }
    
    func Initialize() -> Bool {
        // Clear out all the trips
        DatabaseProvider.ClearAllTrips()
      
        let dbHomeDataArray = (try! Realm()).objects(HomeDataModal.self)
        let homeDataArray: List<HomeDataModal> = List<HomeDataModal>();
      
        for homeData in dbHomeDataArray {
          homeDataArray.append(homeData)
        }
      
        try! self.GenerateHomeData(homeData: homeDataArray)
      
        let photos = PHPhotoLibrary.authorizationStatus()
        if photos == .notDetermined {
          PHPhotoLibrary.requestAuthorization({status in
            if status != .authorized{
              return;
            }
          })
        }
      
        let photoRollInfos: [ClusterModal] = PhotoLibraryProcessor.getPhotosFromLibrary();
      
        // Create a No photos found warning page
        if (photoRollInfos.count == 0) {
          print("Dump: No photos found (Check photo permission?)")
          return true
        }
        
        let trips = try! PhotoLibraryProcessor.GenerateTripFromPhotos(clusterData: photoRollInfos)
        DatabaseProvider.UpdateDBWithTrips(trips: trips)
        return true
    }
  
    
  
  func GenerateHomeData(homeData: List<HomeDataModal>) throws  {
      if(homeData.count == 0) {
        throw EngineError.coreEngineError(message: "No homes as input to be processed")
      }
    
      let today = Date()
    
      // Populate array till today with empty object
      var homesForDataClustering: [HomesForDataClusteringModal] = []
      for _ in 0...Int(today.timeIntervalSince1970/86400) + 1 {
        homesForDataClustering.append(HomesForDataClusteringModal())
      }
    
      // Modify timestamp and location of objects in accordance with homeData input from user
      var currentTimestamp = Date().timeIntervalSince1970/86400
      for data in homeData {
        var coordinates = TripUtils.getCoordinatesFromLocation(location: data.name)
        let _region = coordinates.count > 0 ? coordinates[0] : Region() //Bug, user should select ? 
        
        while(currentTimestamp >= data.timestamp/86400) {
          
          homesForDataClustering[Int(currentTimestamp)].timestamp = currentTimestamp
          homesForDataClustering[Int(currentTimestamp)].name = data.name;
          homesForDataClustering[Int(currentTimestamp)].latitude = _region.latitude;
          homesForDataClustering[Int(currentTimestamp)].longitude = _region.longitude;
          
          currentTimestamp -= 1
        }
      }
    
      DatabaseProvider.UpdateDBWithHomesForDataClustering(homes: homesForDataClustering)
    }
  
  
}

File: android/app/src/main/java/com/beerwithai/glimpse/Engine/Engine.java
package com.beerwithai.glimpse.Engine;

import com.beerwithai.glimpse.Engine.Modals.ClusterModal;
import com.beerwithai.glimpse.Engine.Modals.HomeDataModal;
import com.beerwithai.glimpse.Engine.Modals.HomesForDataClusteringModal;
import com.beerwithai.glimpse.Engine.Modals.Region;
import com.beerwithai.glimpse.Engine.Modals.TripModal;
import com.beerwithai.glimpse.Engine.Providers.DatabaseProvider;
import com.beerwithai.glimpse.Engine.Utils.PhotoLibraryProcessor;
import com.beerwithai.glimpse.Engine.Utils.TripUtils;
import com.facebook.react.bridge.ReactApplicationContext;

import java.util.ArrayList;
import java.util.Date;

public class Engine {
    public static Engine EngineInstance = new Engine();

    public void GenerateHomeData(ArrayList<HomeDataModal> homeData) throws Exception {
        if(homeData.size() == 0) {
            throw new Exception("No homes as input found");
        }

        Date today = new Date();

        ArrayList<HomesForDataClusteringModal> homesForDataClusteringModals = new ArrayList<HomesForDataClusteringModal>();
        for(int i = 0; i < (int)(today.getTime()/86400); i++) {
            homesForDataClusteringModals.add(new HomesForDataClusteringModal());
        }

        int currentTimestamp = (int)(today.getTime()/86400);
        for(int i = 0; i < homeData.size(); i++) {
            try {
                ArrayList<Region> coordinates = TripUtils.getCoordinatesFromLocation(homeData.get(i).name); // TODO: Bug
                Region _region = coordinates.get(0);

                while(currentTimestamp >= homeData.get(i).timestamp/86400) {
                    homesForDataClusteringModals.get((currentTimestamp)).timestamp = (double)currentTimestamp;
                    homesForDataClusteringModals.get((currentTimestamp)).name = homeData.get(i).name;
                    homesForDataClusteringModals.get((currentTimestamp)).latitude = homeData.get(i).latitude;
                    homesForDataClusteringModals.get((currentTimestamp)).longitude = homeData.get(i).longitude;

                    currentTimestamp -= 1;
                }
            } catch (Exception e) {
                // TODO: Handle exception
            }
        }

        DatabaseProvider.UpdateDBWithHomesForDataClustering(homesForDataClusteringModals);
    }

    public boolean Initialize(ReactApplicationContext context) {
        DatabaseProvider.ClearAllTrips(); // Remove this, all logic in react native
        ArrayList<HomeDataModal> homeDataArray = DatabaseProvider.GetHomeData();
        try {
            this.GenerateHomeData(homeDataArray);
        } catch (Exception e) {

        }

        ArrayList<ClusterModal> _cluster = PhotoLibraryProcessor.getPhotosFromLibrary(context);
        _cluster =  PhotoLibraryProcessor.PopulateImageProperties(_cluster);

        if(_cluster.size() == 0) {
            return true;
        }

        try {
            ArrayList<TripModal> trips = PhotoLibraryProcessor.GenerateTripFromPhotos(_cluster);
            DatabaseProvider.UpdateDBWithTrips(trips);
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }
}